src/Microsoft.DotNet.XHarness.Android/AdbRunner.cs

```diff
@@ -335,29 +335,84 @@ public List<string> PullFiles(string devicePath, string localPath)
-                    // if the file is already there, just warn and skip it.
-                    if (File.Exists(destinationPath))
+                    result = RunAdbCommand(new[] { "shell", "run-as", apkPackageName, "ls", devicePath });
+
+                    if (result.ExitCode != (int)AdbExitCodes.SUCCESS)
                     {
-                        _log.LogWarning($"Skipping file copy as {destinationPath} already exists.");
+                        throw new Exception($"Failed checking for file using fallback: {result}");
                     }
-                    else
+
+                    string? fileName = devicePath.Split("/").LastOrDefault();
+                    if (string.IsNullOrWhiteSpace(fileName))
+                    {
+                        throw new Exception($"Failed pulling file using fallback: Couldn't determine filename for {devicePath}");
+                    }
+
+                    string deviceTempPath = $"/data/local/tmp/{fileName}";
+
+                    result = RunAdbCommand(new[] { "shell", "rm", "-f", deviceTempPath });
+
+                    if (result.ExitCode != (int)AdbExitCodes.SUCCESS)
+                    {
+                        throw new Exception($"Failed removing {deviceTempPath} before using fallback: {result}");
+                    }
+
+                    result = RunAdbCommand(new[] { "shell", "touch", deviceTempPath });
+
+                    if (result.ExitCode != (int)AdbExitCodes.SUCCESS)
+                    {
+                        throw new Exception($"Failed touching {deviceTempPath}: {result}");
+                    }
+
+                    result = RunAdbCommand(new[] { "shell", "run-as", apkPackageName, "cp", devicePath, deviceTempPath });
+
+                    if (result.ExitCode != (int)AdbExitCodes.SUCCESS)
+                    {
+                        throw new Exception($"Failed copying file using fallback: {result}");
+                    }
+
+                    result = RunAdbCommand(new[] { "pull", deviceTempPath, tempFolder });
+
+                    if (result.ExitCode != (int)AdbExitCodes.SUCCESS)
+                    {
+                        throw new Exception($"Failed pulling file using fallback: {result}");
+                    }
+
+                    result = RunAdbCommand(new[] { "shell", "rm", "-f", deviceTempPath });
+
+                    if (result.ExitCode != (int)AdbExitCodes.SUCCESS)
                     {
-                        Directory.CreateDirectory(Path.GetDirectoryName(destinationPath) ?? throw new ArgumentException(nameof(destinationPath)));
-                        File.Move(filePath, destinationPath);
-                        copiedFiles.Add(destinationPath);
+                        throw new Exception($"Failed removing {deviceTempPath} after using fallback: {result}");
                     }
                 }
+                else
+                {
+                    throw new Exception($"Failed pulling files: {result}");

```

Comment:
I cannot remember now - do you know if the ADB output is captured in the `result`? In case it is, we would only be displaying exit codes through the all failures to copy the files and I am not sure that might be enough.

Reply:
yes the ADB output is captured in result. It looks like this:

```
fail: Hit error (typically permissions) trying to pull /data/user/0/net.dot.System.Security.Cryptography.Tests/cache/testResults.xml
      System.Exception: Failed pulling file using fallback: Exit code: 1
      Standard Output:
      adb: error: failed to stat remote object '/data/user/0/net.dot.System.Security.Cryptography.Tests/cache/testResults.xml': Permission denied


         at Microsoft.DotNet.XHarness.Android.AdbRunner.PullFiles(String apkPackageName, String devicePath, String localPath) in /Users/alexander/dev/xharness/src/Microsoft.DotNet.XHarness.Android/AdbRunner.cs:line 369
         at Microsoft.DotNet.XHarness.CLI.Commands.Android.AndroidRunCommand.InvokeHelper(ILogger logger, String apkPackageName, String instrumentationName, Dictionary`2 instrumentationArguments, String outputDirectory, String deviceOutputFolder, TimeSpan timeout, Int32 expectedExitCode, WifiStatus wifi, AdbRunner runner) in /Users/alexander/dev/xharness/src/Microsoft.DotNet.XHarness.CLI/Commands/Android/AndroidRunCommand.cs:line 142
```